### PR TITLE
Add BadgeView Control

### PR DESCRIPTION
**Pull request by tnassour**
### Description of Change ###

 

 ### Linked Issues ###
 

 - Fixes #

 ### PR Checklist ###
 
 - [ ] Has a linked Issue, and the Issue has been `approved`(bug) or `Championed` (feature/proposal)
 - [ ] Has tests (if omitted, state reason in description)
 - [ ] Has samples (if omitted, state reason in description)
 - [ ] Rebased on top of `main` at time of PR
 - [ ] Changes adhere to [coding standard](https://github.com/CommunityToolkit/Maui/blob/main/CONTRIBUTING.md#contributing-code---best-practices)
 - [ ] Documentation created or updated: https://github.com/MicrosoftDocs/CommunityToolkit/pulls/XYZ 

 ### Additional information ###

### CODE REVIEW

**Comment by brminnick** [comment]
Assigning [Proposal](https://github.com/CommunityToolkit/Maui/issues/105) Champion, @pictos

File: samples/CommunityToolkit.Maui.Sample/ViewModels/Views/BadgeViewViewModel.cs [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+	{
+		get => counter;
+		set
+		{
+			counter = value;
+			OnPropertyChanged();
+		}
+	}
+
+	public ICommand IncreaseCommand { get; }
```
Comment (VladislavAntonyuk) [inline samples/CommunityToolkit.Maui.Sample/ViewModels/Views/BadgeViewViewModel.cs:27]: Use ICommand attribute instead

File: src/CommunityToolkit.Maui/Views/BadgeView/BadgeView.shared.cs [hunk 2]
```diff
@@ -0,0 +1,469 @@
...
+		BadgeContent.BatchCommit();
+		BadgeIndicatorContainer.BatchCommit();
+		BadgeIndicatorBackground.BatchCommit();
+		BadgeText.BatchCommit();
+		BatchCommit();
+	}
+
+	void UpdateFont()
+	{
+		if (BadgeText == null)
```
Comment (VladislavAntonyuk) [inline src/CommunityToolkit.Maui/Views/BadgeView/BadgeView.shared.cs:328]: is null

File: src/CommunityToolkit.Maui/Views/BadgeView/IBadgeAnimation.shared.cs [hunk 3]
```diff
@@ -0,0 +1,21 @@
+﻿namespace CommunityToolkit.Maui.Views.BadgeView;
+
+/// <summary>
+/// You can implement this interface to create your own animations to be used on the <see cref="BadgeView"/>. Create an implementation of <see cref="IBadgeAnimation"/>, assign the implemenatation to the <see cref="BadgeView.BadgeAnimation"/> and set <see cref="BadgeView.IsAnimated"/> to true.
+/// </summary>
+public interface IBadgeAnimation
```
Comment (VladislavAntonyuk) [inline src/CommunityToolkit.Maui/Views/BadgeView/IBadgeAnimation.shared.cs:6]: It should be in Core

**Review by VladislavAntonyuk — CHANGES_REQUESTED** [review]

**Comment by brminnick** [comment]
Thanks @tnassour! 

I just wanted to give you a heads up that we'll need you to open a PR for our [Docs repo](https://github.com/MicrosoftDocs/CommunityToolkit) before we can merge this PR: https://github.com/MicrosoftDocs/CommunityToolkit

For now, I've added the `pending documentation` and `do not merge` labels so that we don't forget 👍

File: samples/CommunityToolkit.Maui.Sample/CommunityToolkit.Maui.Sample.csproj [hunk 4]
```diff
@@ -65,6 +65,12 @@
     <ProjectReference Include="..\..\src\CommunityToolkit.Maui.SourceGenerators\CommunityToolkit.Maui.SourceGenerators.csproj" OutputItemType="Analyzer" ReferenceOutputAssembly="false" />
   </ItemGroup>
 
+  <ItemGroup>
+    <MauiXaml Update="Pages\Views\BadgeViewPage.xaml">
```
[the remark this answers is not in the document]
Comment (tnassour) [reply samples/CommunityToolkit.Maui.Sample/CommunityToolkit.Maui.Sample.csproj]: Removed

File: samples/CommunityToolkit.Maui.Sample/Pages/Views/BadgeViewPage.xaml [hunk 5]
```diff
@@ -0,0 +1,185 @@
...
+                <Setter Property="FontSize" Value="Medium" />
+                <Setter Property="FontAttributes" Value="Bold" />
+                <Setter Property="Margin" Value="0, 12" />
+            </Style>
+
+        </ResourceDictionary>
+    </pages:BasePage.Resources>
+    <pages:BasePage.Content>
+        <ScrollView>
+            <StackLayout
```
[the remark this answers is not in the document]
Comment (tnassour) [reply samples/CommunityToolkit.Maui.Sample/Pages/Views/BadgeViewPage.xaml]: Changed

File: samples/CommunityToolkit.Maui.Sample/ViewModels/Views/BadgeViewViewModel.cs [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+
+	public BadgeViewViewModel()
+	{
+		Counter = 3;
+
+		IncreaseCommand = new Command(Increase);
+		DecreaseCommand = new Command(Decrease);
+	}
+
+	public int Counter
```
[the remark this answers is not in the document]
Comment (tnassour) [reply samples/CommunityToolkit.Maui.Sample/ViewModels/Views/BadgeViewViewModel.cs]: modified

**Comment by bijington** [comment]
@tnassour is this PR something that you are still able to look at? We are on a feature freeze until .NET MAUI GAs later this month but I would like to see this go into the next version. If you are unable it's no problem, I would be happy to assist in getting it completed.

File: src/CommunityToolkit.Maui/Views/BaseTemplatedView.shared.cs [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+		if (Control != null)
+		{
+			Control.BindingContext = BindingContext;
+		}
+	}
+
+	/// <inheritdoc />
+	protected override void OnChildAdded(Element child)
+	{
+		if (Control == null && child is TControl control)
```
Comment (VladislavAntonyuk) [inline src/CommunityToolkit.Maui/Views/BaseTemplatedView.shared.cs:34]:
```suggestion
		if (Control is null && child is TControl control)
```

**Review by VladislavAntonyuk — CHANGES_REQUESTED** [review]

File: src/CommunityToolkit.Maui/Views/BadgeView/IBadgeAnimation.shared.cs (hunk 3 above)
Comment (tnassour) [reply src/CommunityToolkit.Maui/Views/BadgeView/IBadgeAnimation.shared.cs]: moved

File: src/CommunityToolkit.Maui/Views/BadgeView/BadgeView.shared.cs [hunk 8]
```diff
@@ -0,0 +1,469 @@
+﻿using System.ComponentModel;
+using TypeConverterAttribute = System.ComponentModel.TypeConverterAttribute;
+
+namespace CommunityToolkit.Maui.Views.BadgeView;
+
+/// <summary>
+/// The <see cref="BadgeView"/> allows the user to show a badge with a string value on top of any control. By wrapping a control in a <see cref="BadgeView"/> control, you can show a badge value on top of it. This is very much like the badges you see on the app icons on iOS and Android.
+/// </summary>
+[ContentProperty(nameof(Content))]
+public class BadgeView : BaseTemplatedView<Grid>
```
[the remark this answers is not in the document]
Comment (tnassour) [reply src/CommunityToolkit.Maui/Views/BadgeView/BadgeView.shared.cs]: done

File: src/CommunityToolkit.Maui/Views/BadgeView/BadgeView.shared.cs (hunk 2 above)
Comment (tnassour) [reply src/CommunityToolkit.Maui/Views/BadgeView/BadgeView.shared.cs]: changed

File: src/CommunityToolkit.Maui/Views/BadgeView/BadgeView.shared.cs [hunk 9]
```diff
@@ -0,0 +1,469 @@
...
+	/// Backing BindableProperty for the <see cref="FontFamily"/> property.
+	/// </summary>
+	public static BindableProperty FontFamilyProperty =
+		BindableProperty.Create(nameof(FontFamily), typeof(string), typeof(BadgeView), string.Empty,
+			propertyChanged: OnFontChanged);
+
+	/// <summary>
+	/// Font of the text on the <see cref="BadgeView" />. This is a bindable property.
+	/// </summary>
+	public string FontFamily
```
[the remark this answers is not in the document]
Comment (tnassour) [reply src/CommunityToolkit.Maui/Views/BadgeView/BadgeView.shared.cs:220]: nope cause the label uses both FontFamily, FontAttributes and the Font property is only a getter

File: samples/CommunityToolkit.Maui.Sample/ViewModels/Views/BadgeViewViewModel.cs (hunk 1 above)
Comment (tnassour) [reply samples/CommunityToolkit.Maui.Sample/ViewModels/Views/BadgeViewViewModel.cs]: done

**Comment by tnassour** [comment]
> @tnassour is this PR something that you are still able to look at? We are on a feature freeze until .NET MAUI GAs later this month but I would like to see this go into the next version. If you are unable it's no problem, I would be happy to assist in getting it completed.

@bijington i have done the modifications, it remains testing the sample solution for all of this.
While building locally i am getting this error : 
`3>CSC : error CS0016: Could not write to output file 'C:\Apps\CommunityToolKit\Maui-fork\samples\CommunityToolkit.Maui.Sample\obj\GeneratedFiles\CommunityToolkit.Maui.SourceGenerators\CommunityToolkit.Maui.SourceGenerators.Generators.TextColorToGenerator\SearchBarTextColorTo_9e00a6b2-8703-4c7f-aaa2-32f7268ecea1.g.shared.cs' -- 'Could not find a part of the path 'C:\Apps\CommunityToolKit\Maui-fork\samples\CommunityToolkit.Maui.Sample\obj\GeneratedFiles\CommunityToolkit.Maui.SourceGenerators\CommunityToolkit.Maui.SourceGenerators.Generators.TextColorToGenerator\SearchBarTextColorTo_9e00a6b2-8703-4c7f-aaa2-32f7268ecea1.g.shared.cs'.'`

can someone tell me how to fix this ?

**Comment by VladislavAntonyuk** [comment]
> > @tnassour is this PR something that you are still able to look at? We are on a feature freeze until .NET MAUI GAs later this month but I would like to see this go into the next version. If you are unable it's no problem, I would be happy to assist in getting it completed.
> 
> @bijington i have done the modifications, it remains testing the sample solution for all of this. While building locally i am getting this error : `3>CSC : error CS0016: Could not write to output file 'C:\Apps\CommunityToolKit\Maui-fork\samples\CommunityToolkit.Maui.Sample\obj\GeneratedFiles\CommunityToolkit.Maui.SourceGenerators\CommunityToolkit.Maui.SourceGenerators.Generators.TextColorToGenerator\SearchBarTextColorTo_9e00a6b2-8703-4c7f-aaa2-32f7268ecea1.g.shared.cs' -- 'Could not find a part of the path 'C:\Apps\CommunityToolKit\Maui-fork\samples\CommunityToolkit.Maui.Sample\obj\GeneratedFiles\CommunityToolkit.Maui.SourceGenerators\CommunityToolkit.Maui.SourceGenerators.Generators.TextColorToGenerator\SearchBarTextColorTo_9e00a6b2-8703-4c7f-aaa2-32f7268ecea1.g.shared.cs'.'`
> 
> can someone tell me how to fix this ?

https://github.com/CommunityToolkit/Maui/issues/340#issuecomment-1081052363

File: src/CommunityToolkit.Maui/Views/BaseTemplatedView.shared.cs [hunk 10]
```diff
@@ -0,0 +1,48 @@
+﻿namespace CommunityToolkit.Maui.Views;
+
+/// <summary>
+/// Abstract class that templated views should inherit
+/// </summary>
+/// <typeparam name="TControl">The type of the control that this template will be used for</typeparam>
+public abstract class BaseTemplatedView<TControl> : TemplatedView where TControl : View, new()
```
Comment (VladislavAntonyuk) [inline src/CommunityToolkit.Maui/Views/BaseTemplatedView.shared.cs:7]: TemplatedView uses Compatibility StackLayout. I think we shouldn't use controls from Compatibility namespace

**Comment by brminnick** [comment]
Closing PR as stale + abandoned

**Pull request closed without merging**